kratos/utilities/nurbs_utilities/snake_sbm_utilities.cpp

```diff
@@ -0,0 +1,839 @@
+
+        Vector starting_pos_uv(2);
+        starting_pos_uv[0] = knot_vector_u[0];
+        starting_pos_uv[1] = knot_vector_v[0];
+
+        Vector parameterExternalCoordinates(4);
+        parameterExternalCoordinates[0] = knot_vector_u[0];
+        parameterExternalCoordinates[1] = knot_vector_v[0];
+        parameterExternalCoordinates[2] = knot_vector_u[knot_vector_u.size()-1];
+        parameterExternalCoordinates[3] = knot_vector_v[knot_vector_v.size()-1];
+        
+        surrogate_model_part.GetProcessInfo().SetValue(LOAD_MESHES, parameterExternalCoordinates);
+
+        // Create the matrix of active/inactive knot spans, one for inner and one for outer loop
+        unsigned int numberOfLoops;
+        if (is_inner)
+            numberOfLoops = mParameters["sbm_parameters"]["number_of_inner_loops"].GetInt();
+        else 
+            numberOfLoops = 1;
+        
+        std::vector<int> n_knot_spans_uv(2);
+        n_knot_spans_uv[0] = knot_vector_u.size()-1; 
+        n_knot_spans_uv[1] = knot_vector_v.size()-1;
+
+        std::vector<std::vector<std::vector<int>>> knot_spans_available;
+        knot_spans_available.reserve(numberOfLoops);
+
+        for (IndexType i = 0; i < numberOfLoops; ++i) {
+            std::vector<std::vector<int>> matrix; 
+            matrix.reserve(n_knot_spans_uv[1]);
+            for (int j = 0; j <= n_knot_spans_uv[1]; ++j) {
+                std::vector<int> row(n_knot_spans_uv[0]); 
+                matrix.push_back(row); 
+            }
+            knot_spans_available.push_back(matrix);
+        }
+        
+        // Optimized Snake -> for inner loops
+        int idMatrixKnotSpansAvailable = 0;
+        IndexType idFirstNode;
+        bool newInnerLoop = true;
+
+        
+        KRATOS_INFO_IF("::[SnakeSBMUtilities]::", rEchoLevel > 0 && is_inner) << "Inner :: Starting SnakeStep" << std::endl;
+        KRATOS_INFO_IF("::[SnakeSBMUtilities]::", rEchoLevel > 0 && !is_inner) << "Outer :: Starting SnakeStep" << std::endl;
+                
+        if (skin_model_part_initial.Conditions().size() > 0) {
+            
+            // CREATE FIRST NODE FOR SKIN SUB MODEL PART
+            auto initial_condition = skin_model_part_initial.GetCondition(1);
+            double x_true_boundary0 = initial_condition.GetGeometry()[0].X();
+            double y_true_boundary0 = initial_condition.GetGeometry()[0].Y();
+
+            const int id_first_node = skin_model_part.GetRootModelPart().Nodes().size()+1;
+            skin_sub_model_part.CreateNewNode(id_first_node, x_true_boundary0, y_true_boundary0, 0.0);
+
+            for (auto &i_cond : skin_model_part_initial.Conditions()) {  
+                if (newInnerLoop) {
+                    idFirstNode = i_cond.GetGeometry()[0].Id();
+                    newInnerLoop = false;
+                }
+                // Collect the coordinates of the points of the i_cond
+                double x_true_boundary1 = i_cond.GetGeometry()[0].X();
```

Comment:
`const`?

Reply:
besides, it's better to call `const auto& r_coords i_cond.GtGeometry().[0].Coordinates();` just once and the using `coords[0]` for example